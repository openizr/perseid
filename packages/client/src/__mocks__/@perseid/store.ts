/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * `@perseid/store` mock.
 */

export default class Store {
  public register = vi.fn();

  public mutate = vi.fn();

  public subscribe = vi.fn();

  public combine = vi.fn();

  public dispatch = vi.fn();
}
