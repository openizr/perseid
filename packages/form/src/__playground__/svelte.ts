/* c8 ignore start */
/* eslint-disable @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call */

import { SvelteComponent } from 'svelte';
import Engine from 'scripts/core/Engine';
import Form from 'scripts/svelte/Form.svelte';
import Field from '__playground__/Field.svelte';

let app: Form;
const { log } = console;

function main(): void {
  app = new Form({
    props: {
      Field: Field as typeof SvelteComponent | null | undefined,
      configuration: {
        root: 'root',
        fields: {
          test: { type: 'string', required: true },
        },
        steps: {
          root: {
            fields: ['test'],
          },
        },
        plugins: [
          (engine: Engine): void => {
            engine.on('userAction', (data, next) => {
              log(data);
              return next(data);
            });
          },
        ],
      },
    },
    target: document.getElementById('root') as unknown as HTMLElement,
  });
}

// Ensures DOM is fully loaded before running app's main logic.
// Loading hasn't finished yet...
if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', main);
  // `DOMContentLoaded` has already fired...
} else {
  main();
}

// Ensures subscriptions to Store are correctly cleared when page is left, to prevent "ghost"
// processing, by manually unmounting Vue components tree.
window.addEventListener('beforeunload', () => {
  (app as { $destroy: () => void; }).$destroy();
});
