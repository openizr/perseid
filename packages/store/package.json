{
  "name": "@perseid/store",
  "version": "7.0.2",
  "license": "MIT",
  "type": "module",
  "main": "./core.js",
  "types": "./core.d.ts",
  "description": "Complete state management for both front and back ends",
  "repository": "https://github.com/openizr/perseid",
  "bugs": "https://github.com/openizr/perseid/issues",
  "contributors": [
    {
      "name": "Matthieu Jabbour",
      "email": "matthieu.jabbour@gmail.com",
      "homepage": "https://github.com/matthieujabbour"
    }
  ],
  "keywords": [
    "state",
    "management",
    "react",
    "vue",
    "svelte",
    "store",
    "typescript",
    "javascript",
    "frontend",
    "backend",
    "global",
    "ui",
    "flux",
    "router"
  ],
  "devDependencies": {
    "@perseid/dev-kit": "^8.0.1",
    "@testing-library/react": "^14.0.0",
    "@testing-library/svelte": "^4.0.5",
    "@testing-library/vue": "^8.0.1",
    "@types/react": "^18.2.45",
    "@types/react-dom": "^18.2.18",
    "coveralls": "^3.1.1"
  },
  "eslintConfig": {
    "extends": [
      "./node_modules/@perseid/dev-kit/main.cjs"
    ],
    "overrides": [
      {
        "files": [
          "*.d.ts"
        ],
        "rules": {
          "no-undef": "off",
          "import/order": "off",
          "import/named": "off",
          "import/export": "off",
          "import/no-cycle": "off",
          "react/no-typos": "off",
          "import/extensions": "off",
          "no-use-before-define": "off",
          "import/no-duplicates": "off",
          "import/no-unresolved": "off",
          "max-classes-per-file": "off",
          "import/no-self-import": "off",
          "import/no-named-as-default": "off",
          "import/no-relative-packages": "off",
          "import/no-named-as-default-member": "off",
          "import/no-extraneous-dependencies": "off",
          "@typescript-eslint/no-explicit-any": "off"
        }
      },
      {
        "files": [
          "*.test.ts"
        ],
        "rules": {
          "@typescript-eslint/unbound-method": "off"
        }
      },
      {
        "files": [
          "Store.ts"
        ],
        "rules": {
          "@typescript-eslint/unbound-method": "off",
          "@typescript-eslint/no-dynamic-delete": "off",
          "@typescript-eslint/no-misused-promises": "off"
        }
      },
      {
        "files": [
          "src/scripts/connectors/react.ts"
        ],
        "rules": {
          "react-hooks/exhaustive-deps": "off"
        }
      },
      {
        "files": [
          "src/scripts/connectors/vue.ts"
        ],
        "rules": {
          "@typescript-eslint/no-explicit-any": "off",
          "@typescript-eslint/no-unsafe-assignment": "off"
        }
      }
    ]
  },
  "dependencies": {
    "path-to-regexp": "^6.2.1"
  },
  "engines": {
    "node": ">= 18.0.0",
    "npm": ">= 8.0.0"
  },
  "devKitConfig": {
    "target": "web",
    "devServer": {
      "host": "0.0.0.0",
      "port": "FRONTEND_EXAMPLES_PORT"
    },
    "html": "./__playground__/index.html",
    "entries": {
      "core": "./scripts/core/index.ts",
      "connectors/vue": "./scripts/connectors/vue.ts",
      "connectors/react": "./scripts/connectors/react.ts",
      "connectors/svelte": "./scripts/connectors/svelte.ts",
      "extensions/router": "./scripts/extensions/router.ts"
    },
    "runInDev": true,
    "splitChunks": true,
    "srcPath": "src",
    "distPath": "dist",
    "banner": "/*!\n * Copyright (c) Openizr. All Rights Reserved.\n *\n * This source code is licensed under the MIT license found in the\n * LICENSE file in the root directory of this source tree.\n */"
  },
  "scripts": {
    "test": "node_modules/@perseid/dev-kit/scripts/test",
    "dev": "node node_modules/@perseid/dev-kit/scripts/dev.js",
    "check": "node node_modules/@perseid/dev-kit/scripts/check.js",
    "postinstall": "rm -f node_modules/.eslintcache && mkdir -p node_modules/.cache && test -n \"$POSTINSTALLED\" || POSTINSTALLED=1 yarn add --peer svelte react react-dom vue",
    "build": "cp package.json package.json.back && sed -i s/web/node/g package.json && node node_modules/@perseid/dev-kit/scripts/build.js && yarn run build:types && mv package.json.back package.json",
    "build:types": "cp src/scripts/core.d.ts dist && cp src/scripts/svelte.d.ts dist/connectors && cp src/scripts/vue.d.ts dist/connectors && cp src/scripts/react.d.ts dist/connectors && cp src/scripts/router.d.ts dist/extensions"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "svelte": "^4.2.8",
    "vue": "^3.4.5"
  },
  "peerDependenciesMeta": {
    "react": {
      "optional": true
    },
    "react-dom": {
      "optional": true
    },
    "svelte": {
      "optional": true
    },
    "vue": {
      "optional": true
    }
  }
}