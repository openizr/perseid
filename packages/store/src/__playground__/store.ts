/* c8 ignore start */

import Store from 'scripts/core/Store';
import router from 'scripts/extensions/router';

interface Test {
  test: string;
}

const store = new Store();
store.register('router', router(['/', '/test', '/test2']));
store.subscribe('router', () => {
  // No-op.
});
store.use(() => {
  // No-op.
});
const a: Module<Test> = {
  state: {
    test: 'ok',
  },
  mutations: {
    test({ state }) {
      return { ...state };
    },
  },
  actions: {},
};
store.register('test', a);

store.combine('test2', ['test'], (s1: Test) => ({
  ok: 'test',
  ...s1,
}));

export default store;
