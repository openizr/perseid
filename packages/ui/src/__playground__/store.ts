/* c8 ignore start */

import Store from '@perseid/store';
import router from '@perseid/store/extensions/router';

const store = new Store();
const routes = {
  '/sass': () => import('__playground__/pages/Sass'),
  '/react': () => import('__playground__/pages/Home'),
  '/react/images': () => import('__playground__/pages/Images'),
  '/react/buttons': () => import('__playground__/pages/Buttons'),
  '/react/options': () => import('__playground__/pages/Options'),
  '/react/tooltips': () => import('__playground__/pages/Tooltips'),
  '/react/textareas': () => import('__playground__/pages/Textareas'),
  '/react/typography': () => import('__playground__/pages/Typography'),
  '/react/textfields': () => import('__playground__/pages/Textfields'),
  '/react/file-pickers': () => import('__playground__/pages/FilePickers'),
  '/svelte': () => import('__playground__/pages/Home.svelte'),
  '/svelte/images': () => import('__playground__/pages/Images.svelte'),
  '/svelte/buttons': () => import('__playground__/pages/Buttons.svelte'),
  '/svelte/options': () => import('__playground__/pages/Options.svelte'),
  '/svelte/tooltips': () => import('__playground__/pages/Tooltips.svelte'),
  '/svelte/textareas': () => import('__playground__/pages/Textareas.svelte'),
  '/svelte/typography': () => import('__playground__/pages/Typography.svelte'),
  '/svelte/textfields': () => import('__playground__/pages/Textfields.svelte'),
  '/svelte/file-pickers': () => import('__playground__/pages/FilePickers.svelte'),
  '/vue': () => import('__playground__/pages/HomePage.vue'),
  '/vue/images': () => import('__playground__/pages/ImagesPage.vue'),
  '/vue/buttons': () => import('__playground__/pages/ButtonsPage.vue'),
  '/vue/options': () => import('__playground__/pages/OptionsPage.vue'),
  '/vue/tooltips': () => import('__playground__/pages/TooltipsPage.vue'),
  '/vue/textareas': () => import('__playground__/pages/TextareasPage.vue'),
  '/vue/typography': () => import('__playground__/pages/TypographyPage.vue'),
  '/vue/textfields': () => import('__playground__/pages/TextfieldsPage.vue'),
  '/vue/file-pickers': () => import('__playground__/pages/FilePickersPage.vue'),
} as Record<string, () => Promise<unknown>>;

store.register('router', router(Object.keys(routes)));

export { Store, store, routes };
