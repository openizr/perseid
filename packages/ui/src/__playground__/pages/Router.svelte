/* c8 ignore start */

<script lang="ts">
  import { routes, store } from '__playground__/store';
  import { onDestroy, SvelteComponent } from 'svelte';
  import connect from '@perseid/store/connectors/svelte';
  import type { RoutingContext } from '@perseid/store/extensions/router';

  let component: SvelteComponent | null = null;
  const useSubscription = connect(store);
  const router = useSubscription<RoutingContext>('router');
  const unsubscribe = router.subscribe((state: RoutingContext) => {
    const currentRoute = state.route ?? '';
    if (routes[currentRoute] as unknown !== undefined) {
      routes[currentRoute]().then((response) => {
        component = response as SvelteComponent;
      }).catch(() => null);
    } else {
      component = null;
    }
  });

  onDestroy(unsubscribe);
</script>

{#if component !== null}
  <svelte:component this={component.default} />
{/if}
