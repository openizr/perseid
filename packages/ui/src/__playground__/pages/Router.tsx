/* c8 ignore start */

import * as React from 'react';
import { routes, store } from '__playground__/store';
import connect from '@perseid/store/connectors/react';
import { type RoutingContext } from '@perseid/store/extensions/router';

type Translate = (label: string) => string;
type LazyComponent = () => Promise<{ default: React.ComponentType<{ translate: Translate; }>; }>;

const useCombiner = connect(store);

export default function Router(): JSX.Element {
  const routing = useCombiner<RoutingContext>('router');
  const { route } = routing;

  let currentPage = null;
  if (route !== null && routes[route] as unknown !== undefined) {
    const Component = React.lazy(routes[route] as LazyComponent);
    currentPage = <Component translate={(label: string): string => label} />;
  }

  return (
    <React.Suspense fallback={<div>LOADING...</div>}>
      {currentPage}
    </React.Suspense>
  );
}
