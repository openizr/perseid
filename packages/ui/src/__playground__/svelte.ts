/* c8 ignore start */

import type { SvelteComponent } from 'svelte';
import Router from '__playground__/pages/Router.svelte';

let app: SvelteComponent;

function main(): void {
  const target = document.querySelector('#root') as unknown as HTMLElement;
  target.innerHTML = '';
  app = new Router({
    hydrate: false,
    target,
  });
}

// Ensures DOM is fully loaded before running app's main logic.
// Loading hasn't finished yet...
if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', main);
  // `DOMContentLoaded` has already fired...
} else {
  main();
}

// Ensures subscriptions to Store are correctly cleared when page is left, to prevent "ghost"
// processing, by manually unmounting Svelte components tree.
window.addEventListener('beforeunload', () => {
  app.$destroy();
});
