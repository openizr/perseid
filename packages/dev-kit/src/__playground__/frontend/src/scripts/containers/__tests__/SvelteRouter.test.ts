/**
 * @vitest-environment jsdom
 */

import store from 'scripts/store/index';
import { render, act } from '@testing-library/svelte';
import Router from 'scripts/containers/Router.svelte';

type Misc = any; // eslint-disable-line @typescript-eslint/no-explicit-any

// Useful mocks allowing us to easily test Vue lazy components.
vi.mock('scripts/store/routes', () => ({
  default: {
    // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
    '/': (): Misc => import('scripts/pages/Home.svelte'),
  },
}));

describe('svelte/Router', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders correctly - loading page', () => {
    const { container } = render(Router, { props: { locale: { LABEL_TEST: 'Test' } } });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - page found', async () => {
    store.mutate('router', 'NAVIGATE', '/');
    const { container } = render(Router, { props: { locale: { LABEL_TEST: 'Test' } } });
    await act(() => new Promise((resolve) => {
      setTimeout(resolve, 100);
    }));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - page not found', () => {
    store.mutate('router', 'NAVIGATE', '/404');
    const { container } = render(Router, { props: { locale: { LABEL_TEST: 'Test' } } });
    expect(container.firstChild).toMatchSnapshot();
  });
});
