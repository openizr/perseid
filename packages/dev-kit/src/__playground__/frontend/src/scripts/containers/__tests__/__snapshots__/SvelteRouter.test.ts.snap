// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svelte/Router > renders correctly - loading page 1`] = `
<div>
  <div>
    {"LABEL_TEST":"Test"}
  </div>
  <!--&lt;Loader&gt;-->
  
  
  <!--&lt;Router&gt;-->
</div>
`;

exports[`svelte/Router > renders correctly - page found 1`] = `
<div>
  <p>
    <img
      alt="test"
      src="/images/test.jpg"
    />
     
    <img
      alt="test"
      src="/images/test.svg"
    />
     
    Test
  </p>
  <!--&lt;Home&gt;-->
  
  
  <!--&lt;Router&gt;-->
</div>
`;

exports[`svelte/Router > renders correctly - page not found 1`] = `
<div>
  
  <!--&lt;Router&gt;-->
</div>
`;
